src/main/java/reactor/kafka/sender/ImmutableSenderOptions.java

```diff
@@ -21,51 +21,88 @@
 import reactor.util.concurrent.Queues;
 
 class ImmutableSenderOptions<K, V> implements SenderOptions<K, V> {
 
+    private static final AtomicInteger PRODUCER_CLIENT_ID_SEQUENCE = new AtomicInteger(1);
+
     private final Map<String, Object> properties;
-    private final Serializer<K>       keySerializer;
-    private final Serializer<V>       valueSerializer;
-    private final Duration            closeTimeout;
-    private final Scheduler           scheduler;
-    private final int                 maxInFlight;
-    private final boolean             stopOnError;
+
+    private final Serializer<K> keySerializer;
+
+    private final Serializer<V> valueSerializer;
+
+    private final Duration closeTimeout;
+
+    private final Scheduler scheduler;
+
+    private final int maxInFlight;
+
+    private final boolean stopOnError;
+
+    private final ObservationRegistry observationRegistry;
+
+    private final KafkaSenderObservationConvention observationConvention;
 
     ImmutableSenderOptions() {
         this(new HashMap<>());
     }
 
     ImmutableSenderOptions(Properties properties) {
         this(
-            properties
-                .entrySet()
-                .stream()
-                .collect(Collectors.toMap(
-                    e -> e.getKey().toString(),
-                    Map.Entry::getValue
-                ))
+                properties
+                        .entrySet()
+                        .stream()
+                        .collect(Collectors.toMap(
+                                e -> e.getKey().toString(),
+                                Map.Entry::getValue
+                        ))
         );
     }
 
     ImmutableSenderOptions(Map<String, Object> properties) {
         this.properties = new HashMap<>(properties);
+        maybeOverrideClientId();
         keySerializer = null;
         valueSerializer = null;
 
         closeTimeout = Duration.ofMillis(Long.MAX_VALUE);
         scheduler = Schedulers.immediate();
         maxInFlight = Queues.SMALL_BUFFER_SIZE;
         stopOnError = true;
+        this.observationRegistry = ObservationRegistry.NOOP;
+        this.observationConvention = null;
+    }
+
+    /**
+     * Simulate {@link ProducerConfig#CLIENT_ID_CONFIG} generation.
+     */
+    private void maybeOverrideClientId() {
+        String refinedClientId;
+        boolean userConfiguredClientId = this.properties.containsKey(ProducerConfig.CLIENT_ID_CONFIG);
+        if (userConfiguredClientId) {
+            refinedClientId = (String) this.properties.get(ProducerConfig.CLIENT_ID_CONFIG);
+        } else {
+            String transactionalId = (String) this.properties.get(ProducerConfig.TRANSACTIONAL_ID_CONFIG);
+            refinedClientId =
+                    "producer-" +
+                            (transactionalId != null ? transactionalId : PRODUCER_CLIENT_ID_SEQUENCE.getAndIncrement());
+        }
+        this.properties.put(ProducerConfig.CLIENT_ID_CONFIG, refinedClientId);
```

Comment:
I still have doubts about this thought, but we definitely need some identifier to attach to an `Observation` for traceability.

Reply:
I believe that this is used by kafka producer. So in that case each trace id should be separate for each element or extracted from parent context. When you create SenderOptions it's rather impossible to achieve that, you need to have it on the observation scope when it's being created for each event.